### PR TITLE
Hamming similarity

### PR DESCRIPTION
**Pull request by nbgl**
Implements the hamming similarity for candidate pair generation.

This PR should be functional and reasonably readable.

However, this code is not optimised. Do not expect the same performance as the C++ Dice coefficient code. Optimisation will come later.

(In the future, I will also modify the Python wrapper around the C++ Dice coefficient code to fit better with the new API.)

### CODE REVIEW

File: tests/test_similarities.py [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+
+    assert len(datasets) == 2
+    assert len(datasets[0]) == len(datasets[1]) == recs_per_dataset
+    assert all(len(datasets[0][i]) == length for i in range(recs_per_dataset))
+
+    assert len(indices) == 2
+    assert all(len(i) == len(sims) for i in indices)
+    assert len(candidates) == len(sims)
+    assert not candidates or len(next(iter(candidates))) == 2
+
```
Comment (unzvfu) [inline tests/test_similarities.py]: You should factor out the first halves of these two tests into a separate function.

File: tests/test_similarities.py [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+    for i, j in itertools.product(range(recs_per_dataset), repeat=2):
+        sim = _hamming_sim_f(datasets[0][i], datasets[1][j])
+
+        if sim >= threshold:
+            assert (i, j) in candidates
+            assert candidates[i, j] == sim
+        else:
+            assert (i, j) not in candidates
+
```
Comment (unzvfu) [inline tests/test_similarities.py]: You could conceivably factor out the second halves of the two tests too.

File: anonlink/similarities.py [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+    for records in itertools.product(enumerate(datasets[0]),
+                                     enumerate(datasets[1])):
+        (i0, clk0), (i1, clk1) = records
+        sim = _hamming_sim_f(clk0, clk1)
+
+        if sim >= threshold:
+            c = sim, (i0, i1)
+            candidates[0][i0].append(c)
+            candidates[1][i1].append(c)
+
```
Comment (unzvfu) [inline anonlink/similarities.py]: This for loop is almost exactly the same in the two functions; can't you factor it out somehow?

File: anonlink/similarities.py [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+        sims_arr = np.array(sims, dtype=float)
+        indices_arr = np.array(indices, dtype=int, order='F').T
+    else:
+        sims_arr = np.empty((0,), dtype=float)
+        indices_arr = np.empty((2,0), dtype=int)
+
+    return indices_arr, sims_arr
+
+
+def _hamming_similarity_no_k(
```
Comment (unzvfu) [inline anonlink/similarities.py]: Why not just handle the 'no k' case in the other function with an if statement to skip the `pair_counter` business?

Comment (nbgl) [reply anonlink/similarities.py]: That’s what I had done originally, but I found that it took multiple `if` statements, and having them interspersed through the code made the function look messy. Plus, there are extra optimisations to be made if `k is None`.

File: anonlink/similarities.py (hunk 3 above)
Comment (nbgl) [reply anonlink/similarities.py]: Good catch. Done!

File: tests/test_similarities.py (hunk 2 above)
Comment (nbgl) [reply tests/test_similarities.py]: I think they’re different enough that moving the common bits to a different function would make the code much less clearer.

File: tests/test_similarities.py (hunk 1 above)
Comment (nbgl) [reply tests/test_similarities.py]: Done!

**Pull request merged**